README
Net::Sssssh version 1.000
=========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Socket::MsgHdr (optional)

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2019 by Ton Hospel

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.28.1 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Net::Sssssh.

1.000	2021-06-06 10:05:16 UT
	- original version; created by h2xs 1.23 with options
		-X -n Net::Sssssh


t/bench_timer.pl
#!/usr/bin/perl -w

# Experiment by developer, not part of the distrivution

use strict;
use warnings;

use Devel::Peek;
use Data::Dumper;
$Data::Dumper::Indent   = 1;
$Data::Dumper::Sortkeys = 1;

my $class = shift || "Timer";

my $LOOPS = 1000;
my $SUB_LOOPS = 5;
my $BETWEEN_LOOPS = 2;
my $TIME_RANGE = 3000;
my $MIN_ELEMS = 0;
my $MAX_ELEMS = 30;
my $CHANCE    = 2.1;

my $verbose = 0;
my ($now, @expire);

package Timer;
# Implements a HEAP
use Scalar::Util qw(weaken);
use Carp;

# Timer indices
sub TIME	() { 0 };
sub INDEX	() { 1 };
sub CODE	() { 2 };	# Must come after INDEX

{
my (@timers, @immediate);
# Timers are kept in a simple binary heap @timers
sub new {
    my ($class, $time, $fun) = @_;

    $time += $now;
    my $i = @timers;
    while ($i > 1 && $time < $timers[$i >> 1][TIME]) {
        weaken($timers[$i] = $timers[$i >> 1]);
        $i = ($timers[$i][INDEX] = $i) >> 1;
    }
    my $timer = bless [$time, $i, $fun], $class;
    weaken($timers[$i] = $timer);
    return $timer;
}

sub delete : method {
    my ($timer) = @_;

    my $i = $timer->[INDEX];
    if (!$i) {
        croak "Not a timer reference" unless defined($i) && $i == 0;
        # Could be a timer sitting on the expired queue in run_now
        $#$timer = INDEX if @$timer > INDEX;
        return;
    }
    $timer->[INDEX] = 0;
    # Last element or beyond...
    if ($i >= $#timers) {
        croak "Not a timer reference" if $i > $#timers;
        pop(@timers);
        return;
    }
    my $time = $timers[-1][TIME];
    if ($i > 1 && $time < $timers[$i >> 1][TIME]) {
        # percolate to root
        do {
            weaken($timers[$i] = $timers[$i >> 1]);
            $i = ($timers[$i][INDEX] = $i) >> 1;
        } while ($i > 1 && $time < $timers[$i >> 1][TIME]);
    } else {
        # percolate to leafs
        my $n = @timers-2;
        my $l = $i * 2;
        while ($l < $n) {
            if ($timers[$l][TIME] < $time) {
                if ($timers[$l+1][TIME] < $timers[$l][TIME]) {
                    weaken($timers[$i] = $timers[$l+1]);
                    $timers[$i][INDEX] = $i;
                    $i = $l+1;
                } else {
                    weaken($timers[$i] = $timers[$l]);
                    $timers[$i][INDEX] = $i;
                    $i = $l;
                }
            } elsif ($timers[$l+1][TIME] < $time) {
                weaken($timers[$i] = $timers[$l+1]);
                $timers[$i][INDEX] = $i;
                $i = $l+1;
            } else {
                last;
            }
            $l = $i * 2;
        }
        if ($l == $n && $timers[$l][TIME] < $time) {
            weaken($timers[$i] = $timers[$l]);
            $timers[$i][INDEX] = $i;
            $i = $l;
        }
    }
    weaken($timers[$i] = pop @timers);
    $timers[$i][INDEX] = $i;
}

# Destroy is needed for entries on @immediate
sub DESTROY {
    shift->delete;
}

sub timeout {
    return
        @timers <= 1 ? "inf" :
        $timers[1][0]-$now > 0 ? $timers[1][0]-$now :
        0;
}

sub time : method {
    return shift->[TIME];
}

sub run_now {
    @timers > 1 || @immediate || return;
    # @immediate must be persistent so no timers get lost if a callback dies
    goto EXPIRED if @timers <= 1 ||
        $timers[1][TIME] > $now;
    while (@timers > 2) {
        push @immediate, $timers[1];
        $timers[1][INDEX] = 0;

        my $time = $timers[-1][TIME];
        my $n = @timers-2;
        my $i = 1;
        my $l = 2;
        while ($l < $n) {
            if ($timers[$l][TIME] < $time) {
                if ($timers[$l+1][TIME] < $timers[$l][TIME]) {
                    weaken($timers[$i] = $timers[$l+1]);
                    $timers[$i][INDEX] = $i;
                    $i = $l+1;
                } else {
                    weaken($timers[$i] = $timers[$l]);
                    $timers[$i][INDEX] = $i;
                    $i = $l;
                }
            } elsif ($timers[$l+1][0] < $time) {
                weaken($timers[$i] = $timers[$l+1]);
                $timers[$i][INDEX] = $i;
                $i = $l+1;
            } else {
                last;
            }
            $l = $i * 2;
        }
        if ($l == $n && $timers[$l][TIME] < $time) {
            weaken($timers[$i] = $timers[$l]);
            $timers[$i][INDEX] = $i;
            $i = $l;
        }
        weaken($timers[$i] = pop @timers);
        $timers[$i][INDEX] = $i;
        goto EXPIRED if $timers[1][TIME] > $now;
    }
    if (@timers == 2) {
        $timers[1][INDEX] = 0;
        push @immediate, pop @timers;
    }
  EXPIRED:
    my $fun;
    defined $fun->[CODE] && push @expire, $fun->[CODE] while $fun = shift @immediate;
    # $fun->[CODE] && $fun->[CODE]->() while $fun = shift @immediate;
}
}

package Timer2;
# Naive ARRAY
use Scalar::Util qw(weaken);

# Timer indices
sub TIME	() { 0 };
sub CODE	() { 1 };

{
my @timers;
my $PLACEHOLDER = [0];

sub new {
    my ($class, $time, $fun) = @_;

    $time += $now;
    my $timer = bless [$time, $fun], $class;
    push @timers, $timer;
    weaken($timers[-1]);
    return $timer;
}

sub run_now {
    @timers = sort { ($a //= $PLACEHOLDER)->[TIME] <=> ($b //= $PLACEHOLDER)->[TIME] } @timers;
    my @immediate;
    push @immediate, shift @timers while @timers && ($timers[0] //= $PLACEHOLDER)->[TIME] <= $now;
    my $fun;
    defined $fun->[CODE] && push @expire, $fun->[CODE] while $fun = shift @immediate;
    # $fun->[CODE] && $fun->[CODE]->() while $fun = shift @immediate;

}

sub time : method {
    return shift->[TIME];
}

sub timeout {
    @timers = sort { ($a //= $PLACEHOLDER)->[TIME] <=> ($b //= $PLACEHOLDER)->[TIME] } @timers;
    shift @timers while @timers && $timers[0][TIME] <= 0;
    return
        !@timers ? "inf" :
        $timers[0][TIME]-$now > 0 ? $timers[0][TIME]-$now :
        0;
}

sub delete {
    my $timer = shift;
    $#$timer = TIME;
    $timer->[TIME] = 0;
}

sub DESTROY {
    shift->delete;
}

}

package Timer3;
# Naive double indexed ARRAY
use Scalar::Util qw(weaken);

# Timer indices
sub TIME	() { 0 };
sub CODE	() { 1 };

{
my @timers;
my $PLACEHOLDER = [0];

sub new {
    my ($class, $time, $fun) = @_;

    $time += $now;
    my $timer = bless [$time, $fun], $class;
    push @timers, [$time, $timer];
    weaken($timers[-1][1]);
    return $timer;
}

sub run_now {
    @timers = sort { $a->[TIME] <=> $b->[TIME] } @timers;
    my @immediate;
    push @immediate, shift @timers while @timers && $timers[0][TIME] <= $now;
    my $fun;
    defined $fun->[CODE] && defined $fun->[CODE][CODE] && push @expire, $fun->[CODE][CODE] while $fun = shift @immediate;
    # $fun->[CODE] && $fun->[CODE]->() while $fun = shift @immediate;

}

sub time : method {
    return shift->[TIME];
}

sub timeout {
    @timers = sort { $a->[TIME] <=> $b->[TIME] } grep $_->[CODE], @timers;
    return
        !@timers ? "inf" :
        $timers[0][TIME]-$now > 0 ? $timers[0][TIME]-$now :
        0;
}

sub delete {
    my $timer = shift;
    $#$timer = TIME;
}

sub DESTROY {
    shift->delete;
}

}

package Timer4;
# Naive ARRAY with minimum
use Scalar::Util qw(weaken);

# Timer indices
sub TIME	() { 0 };
sub CODE	() { 1 };

{
my ($min, @timers);
my $PLACEHOLDER = [0];
my $INFINITY = [9**9**9];

sub new {
    my ($class, $time, $fun) = @_;

    $time += $now;
    my $timer = bless [$time, $fun], $class;
    push @timers, $timer;
    weaken($timers[-1]);
    weaken($min = $timer) if defined $min && $time < $min->[TIME] || @timers == 1;
    return $timer;
}

sub run_now {
    my @immediate;
    if (0) {
        my $i = -1;
        my $m = $INFINITY;
        for my $timer (@timers) {
            $timer && $timer->[CODE] || next;
            my $t = $timer;
            if ($timer->[TIME] <= $now) {
                push @immediate, $t;
                # weaken($immediate[-1]);
            } else {
                weaken($timers[++$i] = $t);
                $m = $timer if $timer->[TIME] < $m->[TIME];
            }
        }
        $#timers = $i;
        if ($m == $INFINITY) {
            $min = undef;
        } else {
            weaken($min = $m);
        }
    } else {
        @timers = sort { ($a //= $PLACEHOLDER)->[TIME] <=> ($b //= $PLACEHOLDER)->[TIME] } @timers;
        my $i = 0;
        ++$i while $i < @timers && $timers[$i][TIME] <= $now;
        if ($i) {
            # This inherits weakness ?
            @immediate = splice(@timers, 0, $i);
        }
        if (@timers) {
            weaken($min = $timers[0]);
        } else {
            $min = undef;
        }
    }

    my $fun;
    defined $fun->[CODE] && push @expire, $fun->[CODE] while $fun = shift @immediate;
    # $fun->[CODE] && $fun->[CODE]->() while $fun = shift @immediate;

}

sub time : method {
    return shift->[TIME];
}

sub timeout {
    return if !@timers;

    return $min->[TIME]-$now > 0 ? $min->[TIME]-$now : 0 if $min && $min->[TIME];
    @timers = sort { ($a //= $PLACEHOLDER)->[TIME] <=> ($b //= $PLACEHOLDER)->[TIME] } @timers;
    shift @timers while @timers && $timers[0][TIME] <= 0;
    if (!@timers) {
        $min = undef;
        return "inf";
    }
    weaken($min = $timers[0]);
    return $min->[TIME]-$now > 0 ? $min->[TIME]-$now : 0;
}

sub delete {
    my $timer = shift;
    $#$timer = TIME;
}

sub DESTROY {
    shift->delete;
}

}

package main;
$now = 0;
srand(8);

my %entries;
my $nr = "0";

sub add {
    # Need a scoped lexical to put in the closure
    my $i = ++$nr;
    my $time = int rand($TIME_RANGE);
    $entries{$i} = $class->new($time, $i);
    # $entries{$i} = $class->new($time, sub { push @expire, $i });
    print "Add nr $i with timeout $time\n" if $verbose;
}

sub remove {
    # Delete
    my @nr = sort { $a <=> $b} keys %entries;
    my $i = $nr[rand @nr];
    delete $entries{$i};
    print "Remove nr $i\n" if $verbose;
}

for my $i (1..$LOOPS) {
    for my $j (1..$SUB_LOOPS) {
        # print Dumper(\%entries);
        my $add;
        my $nr_entries = keys %entries;
        if ($nr_entries <= $MIN_ELEMS) {
            $add = 1;
        } elsif ($nr_entries >= $MAX_ELEMS) {
            $add = 0;
        } else {
            $add = int rand($CHANCE);
        }
        if ($add) {
            add();
        } else {
            remove();
        }
    }
    $class->run_now();
    print "Expire @expire\n" if $verbose;
    delete @entries{@expire};
    @expire = ();
    # Simulate the callbacks also adding and removing timers
    for my $i (1..$BETWEEN_LOOPS) {
        add();
        remove();
    }
    ++$now;
    if ($verbose) {
        my @times = sort {$a <=> $b} map $_->time, values %entries;
        print "NOW=$now, TIMES=@times\n";
    }
    my $timeout = $class->timeout;
    print "Timeout=$timeout\n" if $verbose;
}

# Need to clean %entries
# Without this the mere existence of the add and remove subs causes
# class Timer to massively fail during global destruction
%entries = ();


lib/Net/Sssssh/Package.pm
package Net::Sssssh::Package;

# START HISTORY
# autogenerated by release_pm
use strict;
use warnings;
use vars qw($VERSION $SUB_VERSION $release_time %history);
$VERSION = "1.000";
$SUB_VERSION = "074";
$release_time = 1622973916;	## no critic (UselessNoCritic MagicNumbers)
%history = (
  'Changes' => {
    '1.000' => '1.000'
  },
  'MANIFEST' => {
    '1.000' => '1.000'
  },
  'MANIFEST.SKIP' => {
    '1.000' => '1.000'
  },
  'Makefile.PL' => {
    '1.000' => '1.000'
  },
  'README' => {
    '1.000' => '1.000'
  },
  'README.md' => {
    '1.000' => '1.000'
  },
  'bin/sssssh' => {
    '1.000' => '1.000'
  },
  'bin/sssssh_control' => {
    '1.000' => '1.000'
  },
  'bin/udp_nat' => {
    '1.000' => '1.000'
  },
  'bin/udp_recv' => {
    '1.000' => '1.000'
  },
  'bin/udp_send' => {
    '1.000' => '1.000'
  },
  'lib/Net/Sssssh.pm' => {
    '1.000' => '1.000'
  },
  'lib/Net/Sssssh/Package.pm' => {
    '1.000' => '1.000'
  },
  't/00_load.t' => {
    '1.000' => '1.000'
  },
  't/00_syntax.t' => {
    '1.000' => '1.000'
  },
  't/01_syntax.t' => {
    '1.000' => '1.000'
  },
  't/02_fou_encode_decode.t' => {
    '1.000' => '1.000'
  },
  't/03_udp_send.t' => {
    '1.000' => '1.000'
  },
  't/04_udp_recv.t' => {
    '1.000' => '1.000'
  },
  't/lib_extract' => {
    '1.000' => '1.000'
  },
  'version_check' => {
    '1.000' => '1.000'
  }
);

use Carp;

my $epoch_base;

sub SUB_VERSION {
    return $SUB_VERSION;
}

sub FULL_VERSION {
    return "$VERSION.$SUB_VERSION";
}

sub release_time {
    if (!defined $epoch_base) {
	require Time::Local;
	$epoch_base = Time::Local::timegm(0,0,0,1,0,1970);	## no critic (UselessNoCritic MagicNumbers)
    }
    return $release_time + $epoch_base;
}

sub released {
    my ($package, $version) = @_;
    my $p = $package;
    $p =~ s{::}{/}g;
    my $history = $history{"lib/$p.pm"} ||
	croak "Could not find a history for package '$package'";
    my $lowest = 9**9**9;
    for my $v (keys %$history) {
	$lowest = $v if $v >= $version && $v < $lowest;
    }
    croak "No known version '$version' of package '$package'" if
	$lowest == 9**9**9;
    return $history->{$lowest};
}
1;
__END__

=for stopwords Net::Sssssh globals

=head1 NAME

Net::Sssssh::Package - Version and history of Net::Sssssh

=head1 SYNOPSIS

  use Net::Sssssh::Package;

  $epoch_time = Net::Sssssh::Package->release_time();

  $package_version = Net::Sssssh::Package->VERSION;

  $min_package_version = Net::Sssssh::Package::released($module, $module_version);

=head1 DESCRIPTION

In the context of this documentation a C<package> is a set of files that
together make up a perl extension, not to be confused with the more normal
perl concept of a namespace for globals.

A package release with a certain package version number contains a number of
modules and other files with their own version numbers. This module contains a
history of which files with which versions where in which package release and
also knows when the current package was released.

This module contains a few simple methods to query this.

The version number of this Package.pm module is always equal to the version
number of the package release. This means you can use this to query this
release number and also makes it a convenient target for Makefile.PL
dependencies.

=head1 METHODS

=over

=item X<release_time>$epoch_time = Net::Sssssh::Package->release_time()

Returns the the number of non-leap seconds since whatever time the system that
calls this function considers to be the epoch the last time
L<release_pm|release_pm(1)> was run (even if that was on another system with a
different epoch). Since the idea is to run L<release_pm|release_pm(1)> just
before releasing the package this is therefore the value of
L<time()|perlfunc/time> on the system calling this function at the moment the
package was released. This number is suitable for feeding to
L<gmtime()|perlfunc/gmtime> and L<local_time()|perlfunc/local_time>

=item X<VERSION>$package_version = Net::Sssssh::Package->VERSION

This is the normal L<UNIVERSAL::VERSION|UNIVERSAL/VERSION> method you can use
on all modules, but this particular module is guaranteed to have a version
number that is the same as that of the package

=item X<released>$min_package_version = Net::Sssssh::Package::released($module, $module_version)

Given a module name (e.g. Foo::Bar) and a module version number (e.g. 1.023)
returns the lowest package version in which that module (as the file
F<Foo/Bar.pm>) had at least that module version

=back

=head1 EXPORTS

None.

=head1 SEE ALSO

L<release_pm|release_pm(1)> which can be used to keep Package.pm files up to
date.

=cut

# END HISTORY


Makefile.PL
#!/usr/bin/perl -w
# $Id: Makefile.PL 4944 2012-01-04 12:30:35Z hospelt $
use 5.008000;
use strict;
use warnings;
use ExtUtils::MakeMaker;

our $VERSION = '1.000';

# Allows to suppress all program installation with -n (library only)
use Getopt::Std;
our ($opt_n, $opt_y);
getopts("ny") || die "Usage: $0 [-n] [-y]\n";

open(my $fh, "-|", $^X, "t/lib_extract") || die "Could not run 't/lib_extract': $^E";
chomp(my @modules = map "lib/Net/Sssssh/$_.pm", split " ", do { local $/; <$fh>});

my @programs_to_install;

if (!$opt_n || $opt_y) {
    print <<EOT	## no critic (UselessNoCritic InputOutput::RequireCheckedSyscalls)
Note that you can avoid the following questions by passing
the '-n' or '-y' option to 'Makefile.PL'.

EOT
        if !$opt_y;

    # Only do this site_perl substitution for non binary extensions!
    my $site_lib = "$ExtUtils::MakeMaker::Config{siteprefix}/lib/site_perl";
    if ($ExtUtils::MakeMaker::Config{installsitelib} ne $site_lib &&
        grep $_ eq $site_lib, @INC) {
        print <<EOT	## no critic (UselessNoCritic InputOutput::RequireCheckedSyscalls)
Your default perl library install directory is $ExtUtils::MakeMaker::Config{installsitelib}.
This works fine but has the drawback that the library files will not be found
anymore if the perl version gets upgraded. Since $site_lib is in
your default library search path I can use that instead

EOT
            if !$opt_y;
        # arch directories needed to get packlist into /usr/local/lib/site_perl
        map $ExtUtils::MakeMaker::Config{$_} = $site_lib, qw(installsitelib sitelibexp sitearchexp installsitearch) if
            $opt_y ||
            prompt("Use $site_lib as default library install directory ?", "y") =~ /^y/i;
    }

    print <<EOT	## no critic (UselessNoCritic InputOutput::RequireCheckedSyscalls)
This package comes with programs that I can try
to install in $ExtUtils::MakeMaker::Config{installsitebin}.

EOT
        if !$opt_y;
    my %programs = (
        "bin/sssssh"	=> "forward ports",
    );
    for my $program (sort keys %programs) {
        push @programs_to_install, $program if
            $opt_y ||
            prompt("Install $program, a utility to $programs{$program} ?", "y") =~ /^y/i;
    }
}

settings();

WriteMakefile
    (NAME		=> 'Net::Sssssh',
     # finds $VERSION, requires EU::MM from perl >= 5.5
     VERSION_FROM	=> 'lib/Net/Sssssh/Package.pm',
     eval {
         ## no critic (UselessNoCritic MagicNumbers)
         ExtUtils::MakeMaker->VERSION(6.55_01);
         1;
     } ? (BUILD_REQUIRES	=> {
         # To generate README.md when editing source. Not needed for build
         # "Pod::Markdown::Github"	=> "0.02",
         # Only for the tests
         "Test::More"	=> "0.01",
     }) : (),
     PREREQ_PM		=> {
         "Exporter::Tidy"	=> "0.06",
         "Socket::MsgHdr"	=> "0.04",
         # You probably also want a good implementation like Term::ReadLine::Gnu
         "Term::ReadLine"	=> "0.0",
     },
     ABSTRACT		=> 'Simple Socket Server over SSH',
     AUTHOR		=> 'Ton Hospel <Net-Sssssh@ton.iguana.be>',
     $^O eq "MSWin32" ? (
         PM_FILTER	=> '$(PERL) -p -e1',
     ) : (),
     clean		=> {
         FILES => "\$(DISTNAME).ppd ppm cover_db nytprof nytprof.out* md5-versions.old @modules",
     },
     #LICENSE		=> 'perl',
     #Value must be from legacy list of licenses here
     #http://search.cpan.org/perldoc?Module%3A%3ABuild%3A%3AAPI
 );

# START MY
# autogenerated by release_pm
BEGIN {
    package MY;

    use vars qw(%postamble);

    $postamble{ppm} = <<'EOT';
ppm: $(DISTNAME).ppd
	makeppd.pl "--perl=$(PERLRUN)" --min_version=1.016 "--zip=$(ZIP)" "--tar=$(TAR)" "--compress=$(COMPRESS)" --leave=ppm "--objects=$(OBJECT)" $(DISTNAME).ppd $(VERSION)
EOT

    $postamble{ppm3} = <<'EOT';
ppm3: $(DISTNAME).ppd
	makeppd.pl "--perl=$(PERLRUN)" --min_version=1.016 "--zip=$(ZIP)" "--tar=$(TAR)" "--compress=$(COMPRESS)" --ppm_version=3 --leave=ppm3 "--objects=$(OBJECT)" $(DISTNAME).ppd $(VERSION)
EOT

    $postamble{ppm4} = <<'EOT';
ppm4: $(DISTNAME).ppd
	makeppd.pl "--perl=$(PERLRUN)" --min_version=1.016 "--zip=$(ZIP)" "--tar=$(TAR)" "--compress=$(COMPRESS)" --ppm_version=4 --leave=ppm4 "--objects=$(OBJECT)" $(DISTNAME).ppd $(VERSION)
EOT

    $postamble{ppd} = <<'EOT';
$(DISTNAME).ppd: all ppd
EOT

    $postamble{cover} = <<'EOT';
cover:
	cover -delete
	mkdir cover_db
	-HARNESS_PERL_SWITCHES=-MDevel::Cover make test
	cover
EOT

    $postamble{critic} = <<'EOT';
critic:
	perlcritic .
EOT

    $postamble{ppm_install} = <<'EOT';
ppm_install: ppm
	ppm install ppm/$(DISTNAME).ppd
EOT

    $postamble{ppm_uninstall} = <<'EOT';
ppm_uninstall:
	ppm uninstall $(DISTNAME)
EOT

    # You will need to add your own entry for README.md
    $postamble{commit} = <<'EOT';
commit: all README.md
	release_pm
	git commit -av
EOT

    $postamble{version_check} = <<'EOT';
all ::
	$(NOECHO) $(FULLPERL) version_check "$(NAME)" "$(DESTINSTALLSITELIB)" "$(VERSION)" "$(VERSION_FROM)"
EOT

undef &postamble;	## no critic (UselessNoCritic AmpersandSigils)
}
{
    package MY;
    sub postamble {
	return shift->SUPER::postamble() . join("\n", @postamble{sort {uc $a cmp uc $b || $a cmp $b } keys %postamble});
    }
}
# END MY

sub settings {
    package MY;
    our %postamble;

    sub init_PM {
        my $self = shift;
	$self->SUPER::init_PM();
        for my $pm (@modules) {
            $self->{PM}{$pm} = "blib/$pm";
        }
    }

    $postamble{config} = <<"EOT";
config :: @modules

@modules :: bin/sssssh t/lib_extract
	\$(NOECHO) \$(FULLPERL) t/lib_extract \$\@ \$<
EOT

    $postamble{README} = <<'EOT';
README.md: bin/sssssh
	pod2readme -f github $< $@

dist: README.md
EOT
}
